src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -113,20 +113,22 @@ private BoundStatement RewriteEnumeratorForEachStatement(BoundForEachStatement n
                     rewrittenType: elementType,
                     @checked: node.Checked),
                 conversion: node.ElementConversion,
-                rewrittenType: iterationVar.Type,
+                rewrittenType: node.IterationVariableType.Type,
                 @checked: node.Checked);
 
-            // V v = (V)(T)e.Current;
-            BoundStatement iterationVarDecl = MakeLocalDeclaration(forEachSyntax, iterationVar, iterationVarAssignValue);
+            // V v = (V)(T)e.Current;  -OR-  (D1 d1, ...) = (V)(T)e.Current;
+
+            ImmutableArray<LocalSymbol> iterationVariables;
+            BoundStatement iterationVarDecl = LocalOrDeconstructionDeclaration(node, forEachSyntax, iterationVar, iterationVarAssignValue, out iterationVariables);
 
             AddForEachIterationVariableSequencePoint(forEachSyntax, ref iterationVarDecl);
 
             // while (e.MoveNext()) {
-            //     V v = (V)(T)e.Current;
+            //     V v = (V)(T)e.Current;  -OR-  (D1 d1, ...) = (V)(T)e.Current;
             //     /* node.Body */
             // }
 
-            var rewrittenBodyBlock = CreateBlockDeclaringIterationVariable(iterationVar, iterationVarDecl, rewrittenBody, forEachSyntax);
+            var rewrittenBodyBlock = CreateBlockDeclaringIterationVariables(iterationVariables, iterationVarDecl, rewrittenBody, forEachSyntax);
```

Comment:
Need a test to show that new set of variables is created on every iteration. That would be observable if variables are captured in lambdas. Changing variable from one iteration should not affect variables captured in other iterations.